File: datasets/Python/Claude/DataStructures/hard4.py
# Taking inputs manually
n = int(input("Enter N: "))
print("Enter", n, "elements:")
arr = list(map(int, input().split()))

max_si = 0

# For each pair of elements, they can be M1 and M2 in some interval
# We just need to check all pairs
for i in range(n):
    for j in range(i + 1, n):
        # Calculate XOR of the pair
        si = arr[i] ^ arr[j]
        max_si = max(max_si, si)

print("Maximum possible value of Si:", max_si)

File: datasets/C++/Claude/DataStructures/hard4.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int n;
    cout << "Enter N: ";
    cin >> n;
    
    vector<int> arr(n);
    cout << "Enter " << n << " elements: ";
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    
    long long max_si = 0;
    
    // Optimized - check all pairs (O(n^2))
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            long long si = (long long)arr[i] ^ arr[j];
            max_si = max(max_si, si);
        }
    }
    
    cout << "Maximum possible value of Si: " << max_si << endl;
    
    return 0;
}